### PR TITLE
[solc] Standard compiler interface to permit exception leaks and use it in fuzzer

### LINKED ISSUE #8395: [solc] Fuzzing: Leak exceptions
## Abstract

Currently, the solc fuzzers invoke StandardCompiler's compileSolidity API that does **not** leak exceptions, rather "strigifies" them in the json output.

https://github.com/ethereum/solidity/blob/bb09787bd0bcbcbebf7d47584e58363d332f9d92/libsolidity/interface/StandardCompiler.cpp#L763

The fuzzer harness looks for the string "InternalCompilerException" in the json output, and if found, it throws a runtime exception.

https://github.com/ethereum/solidity/blob/b65a165da19f150fbff39eac08729da72e4216c3/test/tools/fuzzer_common.cpp#L71

One issue with this setup is that the fuzzer believes the underlying problem is the line above (and the stack trace leading up to it), and not the original place in solidity code where the ICE was thrown.

This issue tracks support for a version of `compileSolidity` that the fuzzer could use that **leaks** the original exception so that the fuzzer is informed of a given ICE's uniqueless (via stack trace which the fuzzer uses as a proxy to measure uniqueness).

### PR DESCRIPTION
**Pull request by bshastry**
fixes #8395

### CODE REVIEW

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `fd806d5f85528ad1e341e3bc796fac1540ee304d`:
```
Coding style error:
 libsolidity/interface/StandardCompiler.cpp:1306: "Only \"Solidity\" or \"Yul\" is supported as a language.");
 libsolidity/interface/StandardCompiler.cpp:1310: string("JSON logic exception: ") + _exception.what());
 libsolidity/interface/StandardCompiler.cpp:1314: string("JSON runtime exception: ") + _exception.what());
 libsolidity/interface/StandardCompiler.cpp:1318: "Internal exception in StandardCompiler::compile: " +
 libsolidity/interface/StandardCompiler.cpp:1319: boost::diagnostic_information(_exception));
 libsolidity/interface/StandardCompiler.cpp:1323: "Internal exception in StandardCompiler::compile");
 libsolidity/interface/StandardCompiler.cpp:1339: "Only \"Solidity\" or \"Yul\" is supported as a language.");
 libsolidity/interface/StandardCompiler.cpp:1343: string("JSON logic exception: ") + _exception.what());
 libsolidity/interface/StandardCompiler.cpp:1347: string("JSON runtime exception: ") + _exception.what());

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `4047360f8345e604af94bd365add366172f32079`:
```
Coding style error:
 libsolidity/interface/StandardCompiler.cpp:1306: "Only \"Solidity\" or \"Yul\" is supported as a language.");
 libsolidity/interface/StandardCompiler.cpp:1310: string("JSON logic exception: ") + _exception.what());
 libsolidity/interface/StandardCompiler.cpp:1314: string("JSON runtime exception: ") + _exception.what());
 libsolidity/interface/StandardCompiler.cpp:1318: "Internal exception in StandardCompiler::compile: " +
 libsolidity/interface/StandardCompiler.cpp:1319: boost::diagnostic_information(_exception));
 libsolidity/interface/StandardCompiler.cpp:1323: "Internal exception in StandardCompiler::compile");
 libsolidity/interface/StandardCompiler.cpp:1339: "Only \"Solidity\" or \"Yul\" is supported as a language.");
 libsolidity/interface/StandardCompiler.cpp:1343: string("JSON logic exception: ") + _exception.what());
 libsolidity/interface/StandardCompiler.cpp:1347: string("JSON runtime exception: ") + _exception.what());

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `fcc7e915ef34d6035f804615cdd6729a630e4ee5`:
```
Coding style error:
 libsolidity/interface/StandardCompiler.cpp:1306: "Only \"Solidity\" or \"Yul\" is supported as a language.");
 libsolidity/interface/StandardCompiler.cpp:1310: string("JSON logic exception: ") + _exception.what());
 libsolidity/interface/StandardCompiler.cpp:1314: string("JSON runtime exception: ") + _exception.what());
 libsolidity/interface/StandardCompiler.cpp:1318: "Internal exception in StandardCompiler::compile: " +
 libsolidity/interface/StandardCompiler.cpp:1319: boost::diagnostic_information(_exception));
 libsolidity/interface/StandardCompiler.cpp:1323: "Internal exception in StandardCompiler::compile");
 libsolidity/interface/StandardCompiler.cpp:1339: "Only \"Solidity\" or \"Yul\" is supported as a language.");
 libsolidity/interface/StandardCompiler.cpp:1343: string("JSON logic exception: ") + _exception.what());
 libsolidity/interface/StandardCompiler.cpp:1347: string("JSON runtime exception: ") + _exception.what());

```
Please check that your changes are working as intended.

File: test/tools/fuzzer_common.h [hunk 1]
```diff
@@ -24,7 +24,8 @@
 struct FuzzerUtil
 {
 	static void runCompiler(std::string const& _input, bool _quiet);
-	static void testCompiler(std::string const& _input, bool _optimize, bool _quiet);
+	static void testCompilerJsonInterface(std::string const& _input, bool _optimize, bool _quiet);
 	static void testConstantOptimizer(std::string const& _input, bool _quiet);
 	static void testStandardCompiler(std::string const& _input, bool _quiet);
+	static void testCompiler(std::string const& _input, bool _optimize);
```
Comment (leonardoalt) [inline test/tools/fuzzer_common.h:30]: Is this function still used?

**Comment by leonardoalt** [comment]
Compilation error on windows

File: test/tools/fuzzer_common.h (hunk 1 above)
Comment (bshastry) [reply test/tools/fuzzer_common.h:30]: Yes, it is used here

https://github.com/ethereum/solidity/blob/435c9dae718eeef645edb715293c4c687240ee5e/test/tools/afl_fuzzer.cpp#L127

**Comment by bshastry** [comment]
> Compilation error on windows

I think the latest update should fix it.

File: test/tools/fuzzer_common.h (hunk 1 above)
Comment (bshastry) [reply test/tools/fuzzer_common.h:30]: Oops, ignore my previous comment.

`testCompiler` has been renamed `testCompilerJsonInterface`. This is now used exclusively in the afl-fuzzer harness. This API does not leak exceptions and has been retained for backwards compatibility with solfuzzer.

The new harness is called `testCompiler`. This is now used in test files inside ossfuzz sub-directory. This harness leaks exceptions.

**Review by leonardoalt — DISMISSED** [review]

File: test/tools/fuzzer_common.cpp [hunk 2]
```diff
@@ -17,28 +17,87 @@
...
+
+void FuzzerUtil::testCompiler(string const& _input, bool _optimize)
+{
+	frontend::CompilerStack compiler;
+	EVMVersion evmVersion = s_evmVersions[_input.size() % s_evmVersions.size()];
+	frontend::OptimiserSettings optimiserSettings;
+	if (_optimize)
+		optimiserSettings = frontend::OptimiserSettings::standard();
+	else
+		optimiserSettings = frontend::OptimiserSettings::none();
```
Comment (chriseth) [inline test/tools/fuzzer_common.cpp]: Why not minimal?

Comment (bshastry) [reply test/tools/fuzzer_common.cpp]: Ah, I didn't notice it earlier.

https://github.com/ethereum/solidity/blob/d55bbd4aa5dd24ca6f105b24d87127679e1f1810/libsolidity/interface/StandardCompiler.cpp#L454

**Comment by bshastry** [comment]
Updated PR.

File: test/tools/fuzzer_common.cpp [hunk 3]
```diff
@@ -17,28 +17,87 @@
...
-	"tangerineWhistle",
-	"spuriousDragon",
-	"byzantium",
-	"constantinople",
-	"petersburg",
-	"istanbul"
+using namespace solidity::langutil;
+
+static vector<EVMVersion> s_evmVersions = {
+	EVMVersion::homestead(),
```
Comment (chriseth) [inline test/tools/fuzzer_common.cpp:40]: Maybe could be moved to EVMVersion in the future.

**Review by chriseth — APPROVED** [review]

**Pull request merged**